### PR TITLE
JWWEB-722 Send Asset Form steps and validation

### PR DESCRIPTION
**Pull request by flexxnn**

### CODE REVIEW

File: src/pages/Send/Send.js [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+    const sendTransactionPayload: SendTransactionProps = {
+      to: recipientAddress,
+      value: amount,
+      privateKey,
+      gasLimit,
+      gasPrice,
+      // nonce
+    }
+
+    console.log('sendTransactionPayload', sendTransactionPayload)
```
Comment (flexxnn) [inline src/pages/Send/Send.js]: clog

File: src/pages/Send/Send.js [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+  }
+
+  renderSendFormStep = () => {
+    const {
+      sendFormValues,
+    } = this.state
+
+    return (
+      <div className={styles.core}>
+        <TitleHeader
```
[the remark this answers is not in the document]
Comment (flexxnn) [reply src/pages/Send/Send.js]: No, Header should be used when we have right controls inside and don't have back button
When we have back button we should use TitleHeader

File: src/pages/Send/PasswordStepForm.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        values.password,
+        salt,
+      )
+
+      const internalKeyDec = decryptInternalKey(
+        internalKey,
+        derivedKey,
+      )
+
+      const privateKey = getPrivateKey(activeWallet, internalKeyDec)
```
Comment (IViol) [inline src/pages/Send/PasswordStepForm.js]: duplicates `walletsPlugin.getPrivateKey`

File: src/pages/Send/StepOneForm.js [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+    updates: {
+      gasPriceValue: () => this.requestGasPrice(),
+      gasLimitValue: (_, allValues: ?SendFormValues) => allValues
+        ? this.requestGasLimit(allValues)
+        : String(FALLBACK_GAS_AMOUNT),
+    },
+  }, {
+    field: 'recipientAddress',
+    updates: {
+      gasPriceValue: () => this.requestGasPrice(),
```
Comment (IViol) [inline src/pages/Send/StepOneForm.js]: please check that `this.requestGasPrice` is not called for each entered symbol in input

File: src/pages/Send/Send.js [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+    console.log('sendTransactionPayload', sendTransactionPayload)
+
+    try {
+      const txHash = await sendTransaction(
+        network,
+        assetAddress,
+        sendTransactionPayload,
+      )
+
+      console.log('txHash', txHash)
```
Comment (IViol) [inline src/pages/Send/Send.js]: console

File: src/pages/Send/Send.js [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+          contractAddress: null,
+          eventType: checkETH(assetAddress) ? 0 : 1,
+          amount,
+          isRemoved: false,
+        },
+      )
+
+      openTransaction(txHash)
+    } catch (err) {
+      console.error(err)
```
Comment (IViol) [inline src/pages/Send/Send.js]: console

File: src/pages/Send/SendError.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  onCancelClick,
+}: Props) {
+  return (
+    <main
+      className={classNames(
+        '__send_error',
+        style.core,
+      )}
+    >
+      <JIcon
```
Comment (IViol) [inline src/pages/Send/SendError.js]: better to reuse https://github.com/jibrelnetwork/jwallet-web/blob/develop/src/components/UserActionInfo/UserActionInfo.js here

File: src/pages/Send/SendError.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+      <h1 className={style.title}>
+        {t`Internet Connection Error`}
+      </h1>
+      <div className={style.description}>
+        {
+          t`Transfer can't be processed without internet connection. Please, try again later.`
+        }
+      </div>
+      <div className={style.buttons}>
+        <Button
```
Comment (IViol) [inline src/pages/Send/SendError.js]: better to reuse https://github.com/jibrelnetwork/jwallet-web/blob/develop/src/components/ButtonWithConfirm/ButtonWithConfirm.js here

File: src/pages/Send/StepOneForm.js [hunk 9]
```diff
@@ -0,0 +1,388 @@
...
+
+    const {
+      blockchainParams: {
+        decimals,
+      },
+    } = getAssetByAddress(assetAddress)
+
+    const selectedAssetBalance = getAssetBalanceByAddress(assetAddress)
+
+    const amountError = amountValue &&
```
Comment (IViol) [inline src/pages/Send/StepOneForm.js]: better to make it `?string` type

File: src/pages/Send/StepOneForm.js [hunk 10]
```diff
@@ -0,0 +1,388 @@
...
+      await this.requestGasLimit({
+        assetAddress,
+        amountValue,
+        recipientAddress,
+      })
+    }
+
+    return {
+      ...amountError,
+    }
```
Comment (IViol) [inline src/pages/Send/StepOneForm.js]: return amountError && { amountValue: amountError }

File: src/pages/Send/StepOneForm.js [hunk 11]
```diff
@@ -0,0 +1,388 @@
...
+      return ''
+    }
+  }
+
+  requestGasLimit = async ({
+    assetAddress,
+    amountValue,
+    recipientAddress,
+  }: SendFormValues) => {
+    console.log('requestGasLimit', assetAddress, amountValue, recipientAddress)
```
Comment (IViol) [inline src/pages/Send/StepOneForm.js]: console

File: src/pages/Send/StepOneForm.js [hunk 12]
```diff
@@ -0,0 +1,388 @@
...
+        const gasLimit = String(max([requestedGasLimit, staticGasAmount || FALLBACK_GAS_AMOUNT]))
+
+        this.setState({
+          isGasLimitLoading: false,
+        })
+
+        return gasLimit
+      }
+    } catch (err) {
+      console.log('requestGasLimit', err)
```
Comment (IViol) [inline src/pages/Send/StepOneForm.js]: console

File: src/pages/Send/ValidationFailed.js [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+  onGoBackClick: () => any,
+  onGoNextClick: () => any,
+|}
+
+export function ValidationFailed({
+  onGoBackClick,
+  onGoNextClick,
+}: Props) {
+  return (
+    <main
```
Comment (IViol) [inline src/pages/Send/ValidationFailed.js]: UserActionInfo & ButtonWithConfirm could be used here

File: src/pages/Send/components/AssetPicker/AssetPickerContainer.js [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+    )
+    : null
+
+  const assetsWithBalance: DigitalAssetWithBalance[] = getDigitalAssetsWithBalance(
+    activeAssets,
+    assetsBalances,
+  )
+
+  return {
+    fiatCurrency: 'USD',
```
Comment (IViol) [inline src/pages/Send/components/AssetPicker/AssetPickerContainer.js]: `TODO`?

File: src/pages/Send/components/RecipientPicker/RecipientPickerContainer.js [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+      isSimplified,
+      isReadOnly,
+      addressIndex,
+    } = wallet
+
+    const derivationIndex = 0
+
+    if (xpub) {
+      if (isSimplified) {
+        const walletAddresses = generateAddresses(
```
Comment (IViol) [inline src/pages/Send/components/RecipientPicker/RecipientPickerContainer.js]: utils/wallets/getAddresses

File: src/pages/Send/components/RecipientPicker/RecipientPickerContainer.js [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+  +meta: FinalFormMeta,
+  +input: FinalFormInput,
+|}
+
+export function prepareWallets(
+  wallets: Wallet[],
+  walletBalances: Object = {},
+  addressNames: Object = {},
+): RecipientPickerWallet[] {
+  return wallets.map((wallet) => {
```
Comment (IViol) [inline src/pages/Send/components/RecipientPicker/RecipientPickerContainer.js]: map + filter -> reduce?

File: src/pages/Send/send.m.scss [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+.form {
+  @include flex-column-center;
+
+  position: relative;
+  width: 544px;
+}
+
+.mb16 {
+  margin-bottom: 16px;
+}
```
Comment (IViol) [inline src/pages/Send/send.m.scss]: we have styles/offsets for this

File: src/pages/Send/StepOneForm.js (hunk 4 above)
Comment (flexxnn) [reply src/pages/Send/StepOneForm.js]: It works correctly

**Pull request merged**